### PR TITLE
V23 BRANCH: Optimize plotting job memory usage

### PR DESCRIPTION
**Pull request by spxiwh**
THIS WILL BE MERGED ONTO THE v23 RELEASE BRANCH. A SEPARATE PR WILL BE SETUP FOR MERGING THIS
ONTO THE MAIN BRANCH.

We have noticed that many plotting jobs are using a lot of memory. As LDG clusters are now enforcing memory usage more closely, we are having issues getting jobs finished. 6 jobs in particular were identified as offenders:

plot_hist
plot_coinc_snrchi
plot_snrchi
page_foreground
sngls_statmap
merge_page_coincinfo

This patch addresses 5 of them (not sngls_statmap, which is not plotting in any case). 4 of these will also be an issue on the main branch (plot_hist should already be less memory intensive there). Of these 4, 1 is a trivial patch to move across, but others will need reworking to match up with Gareth's changes in #4545. The change to hdf.py also needs Alex to look over it, before merging to the main branch.

Here is performance information before/after this patch for example use cases (page_foreground has a few pathways, so we run 3 examples). The memory usage here is quite low compared to some other examples!

plot_hist:

OLD 	Maximum resident set size (kbytes): 41487604
	User time (seconds): 555.81
NEW 	Maximum resident set size (kbytes): 3135144
	User time (seconds): 115.96
	
plot_coinc_snrchi

OLD	User time (seconds): 173.63
	Maximum resident set size (kbytes): 58355540
NEW	User time (seconds): 181.64
	Maximum resident set size (kbytes): 2447828

plot_snrchi

OLD 	User time (seconds): 989.99
	Maximum resident set size (kbytes): 52556940
NEW	User time (seconds): 227.52
	Maximum resident set size (kbytes): 4724860

page_foreground

OLD1 	User time (seconds): 843.69
	Maximum resident set size (kbytes): 23738796
OLD2	User time (seconds): 168.07
	Maximum resident set size (kbytes): 23715096
OLD3	User time (seconds): 239.99
	Maximum resident set size (kbytes): 23749764

NEW1	User time (seconds): 186.66
	Maximum resident set size (kbytes): 3176904
NEW2	User time (seconds): 44.05
	Maximum resident set size (kbytes): 313200
NEW3	User time (seconds): 124.98
	Maximum resident set size (kbytes): 420688

page_coincinfo

OLD	User time (seconds): 144.31
	Maximum resident set size (kbytes): 23775536
NEW	User time (seconds): 4.79
	Maximum resident set size (kbytes): 207764
	
The outputs of these jobs can be viewed here (apologies, behind an LVK web server, because I had O4 data to hand when doing this ... basically the plots are all identical and XML/HTML lists are identical up to some ordering and rounding).

https://ldas-jobs.ligo.caltech.edu/~ian.harry/pull_requests/plotting_memory_opt/

### CODE REVIEW

File: bin/plotting/pycbc_plot_hist [hunk 1]
```diff
@@ -46,10 +46,31 @@ pycbc.init_logging(args.verbose)
 f = h5py.File(args.trigger_file, 'r')
 ifo = tuple(f.keys())[0]
 
+# This is fixed better on master, but should work here
+if args.x_var == 'snr' or args.x_var == 'newsnr_sgveto':
+    with pycbc.io.HFile(args.trigger_file, 'r') as trig_file:
+        n_triggers_orig = trig_file[f'{ifo}/snr'].size
+        logging.info("Trigger file has %d triggers", n_triggers_orig)
+        logging.info('Generating trigger mask')
+        # psd_var_val may not have been calculated
```
Comment (GarethCabournDavies) [inline bin/plotting/pycbc_plot_hist:55]: as psd_var_val is likely to be used here, can we add something like [this](https://github.com/gwastro/pycbc/blob/v2.3.6/bin/all_sky_search/pycbc_bin_trigger_rates_dq#L57)?

**Review by GarethCabournDavies — APPROVED** [review]
One minor question, but as we don't use psdvar in the plot_hist in the workflow which depends on this change, its probably okay without

File: bin/plotting/pycbc_plot_hist (hunk 1 above)
Comment (spxiwh) [reply bin/plotting/pycbc_plot_hist:55]: I think this is already done on the main branch. We don't need it on the v23 release branch based on current config files (if I missed something let me know!)

Comment (GarethCabournDavies) [reply bin/plotting/pycbc_plot_hist:55]: yup - current configs mean it isn't used, so its not a problem

**Pull request merged**